### PR TITLE
Isolate mlserver dependencies

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/13178?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13178/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13178
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Fix for https://github.com/mlflow-automation/mlflow/actions/runs/10903980766/job/30273563810#step:12:15322

```
  File "/home/runner/work/mlflow/mlflow/tests/openai/mock_openai.py", line 4, in <module>
    import fastapi
  File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/fastapi/__init__.py", line 7, in <module>
    from .applications import FastAPI as FastAPI
  File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/fastapi/applications.py", line 15, in <module>
    from fastapi import routing
  File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/fastapi/routing.py", line 22, in <module>
    from fastapi import params
  File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/fastapi/params.py", line 4, in <module>
    from pydantic.fields import FieldInfo, Undefined
ImportError: cannot import name 'Undefined' from 'pydantic.fields' (/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/pydantic/fields.py)
```

`mlserver` requires an old version of fastapi which is incompatible with `pydantic` v2. This PR isolates mlserver dependencies to avoid fastapi being pined by them.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: dev/pyproject.py [hunk 1]
```diff
@@ -147,6 +142,13 @@ def build(package_type: PackageType) -> None:
                     "boto3>1",
                     "botocore",
                 ],
+                "mlserver": [
+                    # Required to serve models through MLServer
+                    # NOTE: remove the upper version pin once protobuf is no longer pinned in
+                    # mlserver. Reference issue: https://github.com/SeldonIO/MLServer/issues/1089
+                    "mlserver>=1.2.0,!=1.3.1",
+                    "mlserver-mlflow>=1.2.0,!=1.3.1",
```
Comment (harupy) [inline dev/pyproject.py:150]: We might want to drop support for mlserver.

File: mlflow/ml-package-versions.yml [hunk 2]
```diff
@@ -671,6 +671,8 @@ langchain:
           "langchain-core!=0.1.39", # https://github.com/langchain-ai/langchain/issues/19947
           "langchain-openai",
           "langgraph",
+          "fastapi",
+          "uvicorn",
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:675]: `tests/openai/mock_openai.py` requires `fastapi` and `uvicorn`.

File: dev/pyproject.py [hunk 3]
```diff
@@ -147,6 +142,13 @@ def build(package_type: PackageType) -> None:
                     "boto3>1",
                     "botocore",
                 ],
+                "mlserver": [
+                    # Required to serve models through MLServer
+                    # NOTE: remove the upper version pin once protobuf is no longer pinned in
```
Comment (B-Step62) [inline dev/pyproject.py:147]: nit: I think we can remove this note now?

File: mlflow/ml-package-versions.yml (hunk 2 above)
Comment (B-Step62) [reply mlflow/ml-package-versions.yml:675]: The `llama_index` flavor also uses `mock_openai` fixture. Could you add these dependencies there too?

**Review by B-Step62 — APPROVED** [review]
LGTM and agree with splitting `mlserver` from the `[extra]`. 

Can we also update [this step](https://github.com/mlflow/mlflow/blob/4fef77afdbe4d76302cb0b1aad2bd72b5cde64e9/docs/source/deployment/deploy-model-to-kubernetes/tutorial.rst?plain=1#L57-L60) in the k8s deployment tutorial?

File: dev/pyproject.py (hunk 3 above)
Comment (harupy) [reply dev/pyproject.py:147]: ah yeah, let's remove it.

**Comment by klauss42** [comment]
Even if this PR is already closed:

I discovered that the mlserver version pinning is still present in `_install_pyfunc_deps` in `models/container/__init__.py`:
```

    if enable_mlserver:
        server_deps = [
            "'mlserver>=1.2.0,!=1.3.1,<1.4.0'",
            "'mlserver-mlflow>=1.2.0,!=1.3.1,<1.4.0'",
        ]
```

This leads to the issue that an old mlserver version is pulled when using the mlflow generated Dockerfile. Inthe Dockerfile the dependencies are pulled using
```
RUN python -c "from mlflow.models import container as C; C._install_pyfunc_deps('/opt/ml/model', install_mlflow=False, enable_mlserver=True, env_manager='local');"
```

**Pull request merged**